Custom widgets

Augustifolia opened the pull request:
I have started doing some work toward #27, but there is still some more stuff to implement. This far I have renamed some stuff and did some formatting (primarily in WidgetDefinition.py). 

To support setting the valueType of a DirectSpinBox I expanded the use of the loaderFunc. Essentially the type options is stored as strings and the loaderFunc is used to evaluate the string to get the corresponding type. It seams to work quite well, but I am open to ideas about how to improve it.

Augustifolia committed:
Updated custom widget support and fixed DirectSpinBox.widget.

Augustifolia committed:
Fixed exporting of custom loaderFunctions

Augustifolia committed:
Fixed typo in DirectSpinBox.widget

Augustifolia committed:
Added more flexibility to addItemFunction for custom widgets

Augustifolia committed:
Fixed issue with list properties containing data that is not a string

Augustifolia committed:
Fixed pasting of custom elements with a addItemFunction

fireclawthefox commented on DirectGuiDesigner/core/CustomWidgets.py:
```diff
@@ -3,11 +3,13 @@
 
-class CustomWidget():
+
+class CustomWidget:
```
Is there a reason why you removed the brackets? Is that a pep recommendation?

Augustifolia replied:
I mainly think it is a bit ugly to have redundant parentheses, but I don't know if that is a pep recommendation. However I do not really care that much about it

fireclawthefox replied:
Ok, I haven't found much details about this yet but it seems pythons official code also does that so, I'm fine with this change. But, if it's done here, we should also change other parts where applicable to have the code base consistent.

Augustifolia replied:
Yes, there are some other places where this occurs. I will add a commit to change that.

Augustifolia commented on DirectGuiDesigner/tools/JSONTools.py:
The purpose of that change was to make the valueType property of DirectSpinBox show the correct value in the properties panel. Otherwise the string `'float'` would not match the string `float` and as a result `__createOptionMenuProperty` would select the wrong option to show.

Well, I did not find any issues with it when testing. But I might have missed something, so if you would prefer I can try and find some other solution to this?

fireclawthefox replied:
Ah ok, makes sense. I'll have a look at this at a later time too, but for now this should be good then.

Augustifolia replied:
On further thought, I think you where right with this issue. But I think it should be fixed now. Let me know if there are any more problems.

Augustifolia committed:
Removed unnecessary parentheses in some class definitions

Augustifolia committed:
Make it easier to select elements by clicking on them

Done by binding B1PRESS/B1RELEASE to the components of the element as well as the element

Augustifolia committed:
Fixed some edge cases with selecting elements

Augustifolia committed:
Fixed bug with saving position of dragged elements

Sometimes some elements would not save their new position to file if they had been dragged around

Augustifolia committed:
Added ability to specify addItemNode as a list of options

Useful for elements like DirectSplitFrame that has two different frames to reparent to.

Augustifolia committed:
Fixed some issues with loading custom widgets

Made sure the baseWidget is loaded before the child.
Made sure that the baseWidgets definition is inherited even if baseWidget is a custom widget.
Made addItemFunction (and the like) inherit from the baseWidget.

Augustifolia committed:
Added some exception handling to AddItemDialog

Augustifolia committed:
Fixed issue with copying custom widgets

The addItemExtraArgs and addItemNode would not be copied to the new element.

Augustifolia committed:
Added ability to choose an element as parameter to the addItemFunc of custom widgets

fireclawthefox commented on DirectGuiDesigner/dialogs/AddItemDialog.py:
```diff
@@ -95,8 +103,11 @@ def __init__(self, customWidget, child, childInfo, func, rootParent=None):
+        self.radioList = {}  # dict[name: str, list[radiobutton]]
         self.labelList = []
-        self.entryList = []
+        self.entryList = {}
```
If you change the type to dict, please also change the name to represent the change, so it's entryDict. That'll help to know what's in there.

Augustifolia replied:
Yeah, I intended to change that, but I must have forgotten. Thank you for pointing that out

fireclawthefox commented on DirectGuiDesigner/dialogs/AddItemDialog.py:
```diff
@@ -167,35 +181,90 @@ def __callback(self, *args):
+            suppressMouse=True
+        )
+        self.bindScroll(selectionBox)
+        from DirectGuiDesigner.DirectGuiDesigner import DirectGuiDesigner
```
Must this import be here? Of prefer to have imports at the top of the file. If it's just for getting the elementDict, we should better pass out in here somehow. To have better control over who uses this data.

Augustifolia replied:
There is an issue with circular imports if I put if at the top of the file. Perhaps we could add our DirectGuiDesigner instance to base? That way one could use base.DirectGuiDesigner.elementDict instead. 

Otherwise we could just store a reference to DirectGuiDesigner in CusomWidget and AddItemDialog. Or would you prefer that only the elementDict is passed instead of the whole class?

fireclawthefox replied:
Hm, then I'd say passing the elementDict would be the best choice for now except you need more from the DirectGuiDesigner instance itself there. I usually also only passed around the data that was necessary. In general I at least try to not put stuff into global things like base even though I sometimes do it. If passing around the dict would go through too many places to get there, we may should see if the structure can be slimmed down or if not or if it's too complex, then we may put the designer itself into base for easy access.

Augustifolia replied:
Yes, I do generally agree with that. However, in this case the elementDict would have to be passed around a lot to make this happen. We could simplify the structure a bit, but I think that it would still be inconvenient to pass around the elementDict. 

In my opinion, saving DirectGuiDesigner in base would probably be the cleanest solution.

fireclawthefox replied:
Ok, then we may store that data in base. We should start a new task for collecting possible other locations where this can be used then so we don't have both ways all the time but someday only use the base defined value.
Other than that, I think some diagrams for the current dependency structure and what is used where may be helpful if it's getting more and more complex. Will probably have a look at that at a later time, maybe best if you're done with this topic.

Augustifolia committed:
Added reference to DirectGuiDesigner in base

fireclawthefox commented:
That's awesome news. Good job on it so far!

As for point 2, what exactly is the problem that prevents this feature from working? Is it just a displaying thing or more in depth with the data handling?

Regarding point 3, the property panel needs more work anyways. I've started to look at it regarding the performance problems it causes and I think we need a way to pre-generate the texts and other parts and just make it toggle the visibility. So, if your doing larger changes there, let me know and we can work together to make this work much better than it does now.

Number 4 may be an easy fix since we can always change the widgets of the extension itself and fix it there if that's more easy. We may need to provide a more robust implementation of it comes from other widgets though but then it may be a case by case thing.

For point 5, I thought I already had default values but if not, yes, this would definitely be a good addition.

I'll continue to review your changes in the next couple days. From what I've seen so far it does look really good.

Augustifolia commented:
Thank you!

The problem with 2 is that we are basically using elementInfo.type to decide what widget to create and what definition to load. And since they have the same className they will have the same type value (for example "DirectButton"). One solution could be to add a new field with a unique name for each widget type and use that instead of the type.

As for point 3, I do have a few things I would like to change. Stuff like trying to make it more compact to minimize scrolling, being able to edit all values at once in base2/base3/base4 and some other things. But sure, we can work together with those changes. 

I also did some experimentation with loading the properties panel asynchronously, and that actually worked rather well. But your solution will probably work much better than that. 

4: I will take a look. It will hopefully not be to hard to fix.

5: I don't see support for that in the code. But storing that in the definition and just getting it when the element is created shouldn't be hard to do.

fireclawthefox commented:
I see, yes. Though, I'm not quite sure why you would want to have two entries of the same element like DirectButton. Should a new button not be something like CustomDirectButton and use DirectButton as base class? If not, I'm not sure I get what you want to achieve here and you may elaborate a little more in detail or if possible you could give me a little example of why someone would do something like that.

I also had plans to make those base* entries possible to enter just one value for all dimensions, so that sure would be a welcoming addition.
What have you thought about compacting it? That's something that will probably be useful and surely good to have.
As for the asynchronous loading, if you still have that around, we should definitely combine that with the pre-loading of texts to speed things up even more. The more snappy the editor gets, the better, especially since the other editors will probably benefit from it too.

If you need help with Nr. 4, let me know and I'll try to guide you or fix it myself.

Augustifolia commented:
This would be useful for stuff like DirectOptionMenu in DirectGuiExtension. It would also be useful for this project: https://github.com/Augustifolia/BetterDirectGui (written by me). Basically whenever someone makes a copy of some default DirectGui class to make changes to it.

For compacting the properties panel, some ideas are:

1. Put the buttons for changing root parent in a 3x3 grid
2. Whenever there is a button for a property (like some "Browse" or "Add entry" button) put it next to the property header instead of underneath it. That could also be done for bool properties and optionMenu properties.
3. Instead of having a separate label for section headers, just put the header text on the collapsible frame.

Sure, I have the asynchronous stuff here https://github.com/Augustifolia/DirectGuiDesigner/tree/experimental if you want to take a look. It might still need some work thought.

Augustifolia committed:
Added ability to specify default values for properties in widget definition files

Augustifolia committed:
Updated README.md

Augustifolia committed:
Fixed various issues with the DirectGuiExtension integration

Fixed some issues with specific widget types.
Added ability to update some definition of a superclass widget in a .widget file.
Added ability to specify different default values for aspect2d and pixel2d mode in a .widget file.

Augustifolia commented:
I think most issues should be resolved now. 

I have not fixed 2 and 3. I do plan on fixing 3, but that is a bit outside the scope of this PR.

fireclawthefox commented:
Hey, sorry for being so late with the review and merge, I didn't had any time to spare from work and family to check this. But now I finally had some time to check it and yes, it all looks good.

Again, thank you so much for your fantastic contribution!

Augustifolia commented:
No worries, work and family comes first. I will probably start working on some stuff with the properties panel in the near future, but I am also rather busy right now so it might take a couple of weeks before I start.

You're welcome!
